Typewriter 2.0 [DOC-557] (#3253)
segmentio/segment-docs: src/protocols/apis-and-extensions/typewriter.md

![Order Completed Event in a Protocols Tracking Plan](images/typewriter-order-completed.png)

> Typewriter uses the event to generate a typed analytics call in different languages:

Reviewer: I like this edit, but it makes the statement on line 8 incomplete.

## Browser Quickstart

To get started with Typewriter in your browser:
1. Make sure you have `node` installed using the instructions in the [prerequisites](#prerequisites) above.
2. Install `analytics.js` in your app. For now, you just need to complete [`Step 1: Copy the Snippet`](/docs/connections/sources/catalog/libraries/website/javascript/quickstart/#step-2-copy-the-segment-snippet) from the [`analytics.js` Quickstart Guide](/docs/connections/sources/catalog/libraries/website/javascript/quickstart/).
3. Once you've got `analytics.js` installed, add Typewriter as a developer dependency in your project:

    ```sh
    $ npm install --save-dev typewriter
    ```

> 4. Run `npx typewriter init` to use the Typewriter quickstart wizard that generates a [`typewriter.yml`](#configuration-reference) configuration, along with your first Typewriter client. When you run the command, it creates a `typewriter.yml` file in your repo. For more information on the format of this file, see the [Typewriter Configuration Reference](#configuration-reference).

Author: @oscb is this correct?

})
```

> To help you minimize your bundle size, Typewriter supports [tree-shaking](https://webpack.js.org/guides/tree-shaking/){:target="_blank"} using named exports. All generated analytics calls are automatically directly exported, so you can import them like so:

Reviewer: ```suggestion
To help you minimize your bundle size, Typewriter supports [tree-shaking](https://webpack.js.org/guides/tree-shaking/){:target="_blank"} using named exports. All generated analytics calls generate and export automatically, so you can import them like so:
```
Just a thought, I think you can make the sentence more active and remove an adjective.

})
```

> Typewriter wraps your analytics calls in an [ES6 `Proxy`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy){:target="_blank"}, which helps protect your application from crashing if you make analytics calls with a generated function that doesn't exist. For example, if an `Order Completed` event didn't exist in your Tracking Plan in the first example above, then your app would crash with a `TypeError: typewriter.orderCompleted is not a function`. However, since `typewriter` dynamically proxies the underlying function calls, it can detect if a function doesn't exist, and handle it for you. Typewriter logs a warning message, then fires an `Unknown Analytics Call Fired` event into your source. Segment has found this useful when migrating JavaScript projects to Typewriter in bulk, since it gives Segment confidence that Segment won't introduce regressions that crash the application. Keep in mind that proxying does not work with named exports.

Reviewer: ```suggestion
Typewriter wraps your analytics calls in an [ES6 `Proxy`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy){:target="_blank"}, which helps protect your application from crashing if you make analytics calls with a generated function that doesn't exist. For example, if an `Order Completed` event didn't exist in your Tracking Plan in the first example above, then your app would crash with a `TypeError: typewriter.orderCompleted is not a function`. However, since `typewriter` dynamically proxies the underlying function calls, it can detect if a function doesn't exist, and handle it for you. Typewriter logs a warning message, then fires an `Unknown Analytics Call Fired` event into your source. This is useful when you migrate JavaScript projects to Typewriter in bulk, since it gives Segment confidence that Segment won't introduce regressions that crash the application. Keep in mind that proxying does not work with named exports.
```

"...it gives Segment confidence that Segment won't..." Is this correct? It reads as confusing to me that Segment needs to give it self confidence.

})
```

> Typewriter wraps your analytics calls in an [ES6 `Proxy`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy), which helps protect your application from crashing if you make analytics calls with a generated function that doesn't exist. For example, if an `Order Completed` event didn't exist in your Tracking Plan in the first example above, then your app would crash with a `TypeError: typewriter.orderCompleted is not a function`. However, since `typewriter` dynamically proxies the underlying function calls, it can detect if a function does not exist, and handle it for you. Typewriter logs a warning message, then fires an `Unknown Analytics Call Fired` event into your source. Segment has found this useful when migrating JavaScript projects to Typewriter in bulk, since it gives Segment confidence that Segment won't introduce regressions that crash the application. Keep in mind that proxying does not work with named exports.

Reviewer: Same as above

## Swift Quickstart

To get started using Typewriter with Swift:
1. Make sure you have `node` installed using the instructions in the [prerequisites](#prerequisites) above.
2. Install `analytics-swift` in your app. Follow the [analytics-swift Quickstart Guide](/docs/connections/sources/catalog/libraries/mobile/swift-ios).
3. Typewriter comes with a quickstart wizard that generates a [`typewriter.yml`](#configuration-reference) configuration, along with your first Typewriter client. Running the command creates a `typewriter.yml` file in your repo. For more information on the format of this file, see the [Typewriter Configuration Reference](#configuration-reference). To use this wizard, run:

```sh
$ npx typewriter init
```

> info ""
> You can regenerate your Typewriter client by running `npx typewriter`. You need to do this each time you update your Tracking Plan.

4. Import your new Typewriter client into your project using XCode. If you place your generated files into a folder in your project, import the project as a group not a folder reference.

> After adding the generated client to your Xcode Project you will be able to use it from Swift as extension methods on any Analytics client object:

Reviewer: ```suggestion
When you add the generated client to your Xcode Project you can use as a Swift extension method on any Analytics client object:
```

Active voice.

)
```

> info ""
> For use with the `analytics-ios` SDK, use the previous version of [Typewriter v7](https://www.npmjs.com/package/typewriter/v/7.4.1){:target="_blank"}.

## Kotlin Quickstart
To get started using Typewriter with Kotlin:
1. Make sure you have `node` installed. Use the instructions in the [prerequisites](#prerequisites) above.
2. Install `analytics-kotlin` in your app. Follow the [analytics-kotlin QuickStart Guide](/docs/connections/sources/catalog/libraries/mobile/kotlin-android/#getting-started).

> 3. Typewriter comes with a quickstart wizard that generates a [`typewriter.yml`](#configuration-reference) configuration, along with your first Typewriter client. Running the command creates a `typewriter.yml` file in your repo. For more information on the format of this file, see the [Typewriter Configuration Reference](#configuration-reference). To use this wizard, run:

Reviewer: ```suggestion
3. Typewriter comes with a quickstart wizard that generates a [`typewriter.yml`](#configuration-reference) configuration, along with your first Typewriter client. The command creates a `typewriter.yml` file in your repo. For more information on the format of this file, see the [Typewriter Configuration Reference](#configuration-reference). To use this wizard, run:
```
It might also read more clearly to list the command first, since you reference it in the step, and then explain what it does.

Author: ```suggestion
3. Run `npx typewriter init`. This command enables you to use the Typewriter quickstart wizard that generates a [`typewriter.yml`](#configuration-reference) configuration, along with your first Typewriter client. The command creates a `typewriter.yml` file in your repo. For more information on the format of this file, see the [Typewriter Configuration Reference](#configuration-reference).
```

Author: @oscb how's this 👆

> note ""
> You can regenerate your Typewriter client by running `npx typewriter`. You need to do this each time you update your Tracking Plan.

> By default Typewriter creates the class file with a package name of `typewriter`. Segment recommends you to set up the right package name during `npx typewriter init` by choosing to review the additional options for Kotlin. You can also set up the right package name directly in `typewriter.yml`:

Reviewer: ```suggestion
Typewriter creates the class file with a package name of `typewriter`. Segment recommends you to set up the right package name during `npx typewriter init` by choosing to review the additional options for Kotlin. You can also set up the right package name directly in `typewriter.yml`:
```
Nitpicking a bit here, is the package name something that someone simply enters in the config wizard? If so, it might be more clear to say `Segment recommends that you enter the correct package name...` rather than `set up the right package name`

Author: ```suggestion
Typewriter creates the class file with the package name `typewriter`. Segment recommends you to enter the right package name during `npx typewriter init` by choosing to review the Advanced Options for Kotlin. You can also set up the right package name directly in `typewriter.yml`:
```

Author: how's this @oscb

);
```

## React Native Quickstart

To get started with React Native:
1. Follow the [Getting Started guide for React Native](https://segment.com/docs/connections/sources/catalog/libraries/mobile/react-native/).
2. Add `typewriter` as a dev dependency in your project once you have the library installed in your project.

```
$ npm install --save-dev typewriter
```

3. Typewriter comes with a quickstart wizard that generates a typewriter.yml configuration, along with your first Typewriter client. To use this wizard, run:

```
$ npx typewriter init
```

> You can regenerate your Typewriter client by running `npx typewriter`. You need to do this each time you update your Tracking Plan. Running the command creates a `typewriter.yml` file in your repo. For more information on the format of this file, see the [Typewriter Configuration Reference](#configuration-reference). The command also adds a new Typewriter / Segment client in `./analytics` (or whichever path you configured). You can use this interchangeably as a normal React Native Segment client. It contains additional methods for your tracking plan:

Reviewer: ```suggestion
Run `npx typewriter` to regenerate your Typewriter client. You need to do this each time you update your Tracking Plan. This command creates a `typewriter.yml` file in your repo. For more information on the format of this file, see the [Typewriter Configuration Reference](#configuration-reference). The command also adds a new Typewriter / Segment client in `./analytics` (or whichever path you configured). You can use this interchangeably as a normal React Native Segment client. It contains additional methods for your tracking plan:
```

Active voice

## API Token Configuration

> Typewriter requires a Segment API token to fetch Tracking Plans from the [Segment Public API](https://api.segmentapis.com/docs/).

Reviewer: Does this update work only with PAPI? If not, we should list both options, and explain that PAPI is in beta.

> info ""

> > If you're using the previous version of Typewriter that fetches tracking plans from the [Segment Config API](/docs/config-api/) and would like to upgrade to the new version of Typewriter, generate a new API token by following the steps below.

Reviewer: ```suggestion
> If you use a previous version of Typewriter that fetches tracking plans from the [Segment Config API](/docs/config-api/) and would like to upgrade to the new version of Typewriter, generate a new API token by following the steps below.
```

## Token Script

> warning ""
> Segment is keeping the Token Script execution for compatibility purposes only in v8 of Typewriter. Segment might deprecate this feature in the future, and encourages you to execute your script and pipe in the token. For example, `echo $TW_TOKEN | typewriter build`.

If your team has a standard way to supply secrets (passwords and tokens) in development environments, whether that's an `.env` file or an AWS-backed secret store, you can configure Typewriter to use it to get a Segment API token.

> You configure this by creating a token script called `scripts.token` in your `typewriter.yml`. This script is a string that contains a shell command that, when executed, outputs a valid Segment API token. Here's an **insecure**, example:

Reviewer: ```suggestion
To configure this, create a token script called `scripts.token` in your `typewriter.yml`. This script is a string that contains a shell command that, when executed, outputs a valid Segment API token. Here's an **insecure**, example:
```

  onViolation: yourViolationHandler
})
```

> > into ""

Author: ```suggestion
> info ""
```

## Known Limitations

Typewriter only supports `track` calls. However, you can continue to use the underlying (untyped) analytics instance to perform `identify`, `group`, `page`, `screen`, and `alias` calls.

> Not all languages support run-time validation. Currently, `analytics.js` and `analytics-node` support it using [AJV](https://github.com/epoberezkin/ajv) (both for JavaScript and TypeScript projects) while `analytics-ios` and `analytics-android` do not yet support run-time validation. Typewriter also does not yet support run-time validation using Common JSON Schema.

Author: @oscb does Swift and Kotlin support run-time validation?

Reviewer 2: No, they don't at the moment. The comment is still valid: only analytics.js and analytics-node have runtime validation

Author: Then should I replace analytics-ios and analytics-android with analytics-swift and analytics-kotlin?

Reviewer 2: Ah yes, good point! let's replace those with swift and kotlin

> Not all languages support run-time validation. Currently, `analytics.js` and `analytics-node` support it using [AJV](https://github.com/epoberezkin/ajv) (both for JavaScript and TypeScript projects) while `analytics-ios` and `analytics-android` do not yet support run-time validation. Typewriter also does not yet support run-time validation using Common JSON Schema.

Author: ```suggestion
Not all languages support run-time validation. Currently, `analytics.js` and `analytics-node` support it using [AJV](https://github.com/epoberezkin/ajv){:target="_blank"} (both for JavaScript and TypeScript projects) while `analytics-swift` and `analytics-kotlin` don't support run-time validation. Typewriter also doesn't support run-time validation using Common JSON Schema.
```

## Known Limitations

> Typewriter currently only supports `track` calls. However, you can continue to use the underlying (untyped) analytics instance to perform `identify`, `group`, `page`, `screen`, and `alias` calls.

Reviewer: ```suggestion
Typewriter supports `track` calls. However, you can continue to use the underlying (untyped) analytics instance to perform `identify`, `group`, `page`, `screen`, and `alias` calls.
```

Not all languages support run-time validation. Currently, `analytics.js` and `analytics-node` support it using [AJV](https://github.com/epoberezkin/ajv) (both for JavaScript and TypeScript projects) while `analytics-ios` and `analytics-android` do not yet support run-time validation. Typewriter also does not yet support run-time validation using Common JSON Schema.

## Contributing

If you're interested in contributing, [open an issue on GitHub](https://github.com/segmentio/typewriter/issues/new) and Segment can help provide you pointers to get started!

## Feedback

> Segment is always curious about any feedback you have on your experience with Typewriter! To contact Segment, [open an issue on GitHub](https://github.com/segmentio/typewriter/issues/new).

Author: ```suggestion
Segment welcomes feedback you may have on your experience with Typewriter. To contact Segment, [open an issue on GitHub](https://github.com/segmentio/typewriter/issues/new).
```

> Segment is always curious about any feedback you have on your experience with Typewriter! To contact Segment, [open an issue on GitHub](https://github.com/segmentio/typewriter/issues/new).

Reviewer: Can you rewrite this to get rid of the `!`?

"Segment welcomes feedback you may have on your experience with Typewriter." or something.
